metric/internal/global/meter_test.go

```diff
@@ -0,0 +1,265 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package global // import "go.opentelemetry.io/otel/metric/internal/global"
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/metric/instrument/asyncfloat64"
+	"go.opentelemetry.io/otel/metric/instrument/syncfloat64"
+	"go.opentelemetry.io/otel/metric/nonrecording"
+)
+
+func TestMeterProviderRace(t *testing.T) {
+	mp := &meterProvider{}
+	finish := make(chan struct{})
+	go func() {
+		for i := 0; ; i++ {
+			mp.Meter(fmt.Sprintf("a%d", i))
+			select {
+			case <-finish:
+				return
+			default:
+			}
+		}
+	}()
+
+	mp.setDelegate(nonrecording.NewNoopMeterProvider())
+	close(finish)
+
+}
+
+func TestMeterRace(t *testing.T) {
+	mtr := &meter{}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	finish := make(chan struct{})
+	go func() {
+		for i, once := 0, false; ; i++ {
+			name := fmt.Sprintf("a%d", i)
+			_, _ = mtr.AsyncFloat64().Counter(name)
+			_, _ = mtr.AsyncFloat64().UpDownCounter(name)
+			_, _ = mtr.AsyncFloat64().Gauge(name)
+			_, _ = mtr.AsyncInt64().Counter(name)
+			_, _ = mtr.AsyncInt64().UpDownCounter(name)
+			_, _ = mtr.AsyncInt64().Gauge(name)
+			_, _ = mtr.SyncFloat64().Counter(name)
+			_, _ = mtr.SyncFloat64().UpDownCounter(name)
+			_, _ = mtr.SyncFloat64().Histogram(name)
+			_, _ = mtr.SyncInt64().Counter(name)
+			_, _ = mtr.SyncInt64().UpDownCounter(name)
+			_, _ = mtr.SyncInt64().Histogram(name)
+			_ = mtr.RegisterCallback(nil, func(ctx context.Context) {})
+			if !once {
+				wg.Done()
+				once = true
+			}
+			select {
+			case <-finish:
+				return
+			default:
+			}
+		}
+	}()
+
+	// wg.Wait()
```

Comment:
Is the wait group needed anymore?

Reply:
That was a mistake.  If we don't have instruments registered once, which is why the wait group is used, before we `setDelegate` then the race detector doesn't catch this.  I'm not 100% sure why, but this is the case even if there is only 1 call within the for loop.